Retry to get a connection in JDBC connectors

takezoe opened the pull request:
JDBC connector consumes one JDBC connection per table scan. So, if a single query contains tens or hundreds of table scans, it can hit the max connection limit in the database side. In this case, the connection pooling usually can be a solution, however, in Presto, there are various of worker instances in a single cluster, so simply introducing the connection pooling doesn't work.

Retrying to get JDBC connection can make JDBC connector more robust. It would work as an alternative to the connection pooling.

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DriverConnectionFactory.java:
```diff
@@ -60,9 +65,30 @@ public Connection openConnection(JdbcIdentity identity)
+        int retryCount = 0;
+        SQLException error = null;
+
+        while (retryCount <= 10) {
```
Can we make it configurable ?

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DriverConnectionFactory.java:
```diff
@@ -60,9 +65,30 @@ public Connection openConnection(JdbcIdentity identity)
+                return connection;
+            }
+            catch (SQLException e) {
+                retryCount = retryCount + 1;
```
Instead of having it here can we wrap the `DriverConnectionFactory' with a  `RetryableDriverConnectionFactory` which could capture the exception and retry it for n times ? We can configure the initial retry time and the retry count in it ?

takezoe replied:
@Praveen2112 I thought we can add properties to `BaseJdbcConfig` for making the retry behavior configurable. However, if we will create `RetryableDriverConnectionFactory` aside from `DriverConnectionFactory`, it may not be a good idea?

Also, should all JDBC connectors be updated to use `RetryableDriverConnectionFactory` if it will be available in base-jdbc module?

electrum commented:
Consider using https://jodah.net/failsafe/retry/ which we use elsewhere in Presto. This allows a combination of max attempts and max time, with exponential back off.

takezoe commented:
Thanks for your review! I will update this pull request.

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/DriverConnectionFactory.java:
```diff
@@ -57,12 +60,19 @@ public DriverConnectionFactory(Driver driver, String connectionUrl, Properties c
-        checkState(connection != null, "Driver returned null connection");
-        return connection;
+
+        RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
```
Can we initialise it once ?

findepi replied:
it's mutable

takezoe commented:
Created `RetrybleConnectionFactory` as a wrapper of `ConnectionFactory` and applied it to mysql, postgresql and sqlserver plugins.

Now I'm wondering how `RetrybleConnectionFactory` can be configurable.

tooptoop4 commented:
@takezoe is this similar to https://github.com/SlicingDice/presto/commit/3825270f32afb9c91e45c809a911e89a8172275a ?

takezoe commented:
Yes, but that looks not configurable.

takezoe commented:
I just made `RetryableConnectionFactory`, but it can be configured through only constructor parameters.

takezoe commented:
@Praveen2112 @electrum @findepi Does this have a chance to be adopted? and is there something missing for that?

Praveen2112 commented:
@takezoe Sorry for the delay. Instead of wrapping `RetryableConnectionFactory` for each client, can we try like `StatisticsAwareConnectionFactory`

takezoe commented:
@Praveen2112 I'm trying it but I don't fully understand how actually we can use `StatisticsAwareConnectionFactory`. To use it, we need to create our own JDBC plugin?

takezoe commented:
@Praveen2112 Updated my pull request to wrap `RetryableDriverConnectionFactory` by `StatisticsAwareDriverConnecfionFactory` using annotated component binding. Does this make sense?

Praveen2112 commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,84 @@
+        // Max retries = 3 by default
+        ConnectionFactory connectionFactory = new RetryableConnectionFactory(delegate);
+
+        Connection conn = connectionFactory.openConnection(new JdbcIdentity("test", Optional.empty(), Collections.emptyMap()));
```
Using `ImmutableMap.of()`

takezoe replied:
Fixed.

Praveen2112 commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,84 @@
+import static org.testng.Assert.assertNull;
+
+public class TestRetryableConnectionFactory
+{
```
It would be helpful if we add test to ensure all methods are overriden
```
    @Test
    public void testEverythingImplemented()
    {
        assertAllMethodsOverridden(ConnectionFactory.class, StatisticsAwareConnectionFactory.class);
    }
```

takezoe replied:
`RetryableConnectionFactory` doesn't override `close()`.

Praveen2112 replied:
Any reason for not overriding them ? Close API should also be transferred to the delegate right ?

takezoe replied:
Ah, yeah. It must be overriden.

takezoe commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,71 @@
+    @Inject
+    public RetryableConnectionFactory(@ForBaseJdbc ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(1), Duration.ofMinutes(1), 10, 2.0, 0.2);
```
Added `RetryJdbcConfig`.

takezoe commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,84 @@
+                throws SQLException
+        {
+            counter++;
+            System.out.println(counter);
```
Removed. 🙇

takezoe commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,71 @@
+
+        requireNonNull(delay);
+        this.delay = delay.toMillis();
+        System.out.println(this.delay);
```
Removed. 🙇

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,70 @@
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        requireNonNull(delay);
+        this.delay = delay.toMillis();
```
Can we inline them with the previous statement ?

takezoe replied:
Fixed.

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,90 @@
+
+import java.time.Duration;
+
+public class RetryJdbcConfig
```
Can we have a test for this Config ?

takezoe replied:
Added `TestRetryJdbcConfig`

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,90 @@
+        return this;
+    }
+
+    @NotNull
```
It would be great if we could have some validation like Min value to be set

takezoe replied:
Added Min and MinDuration annotations

Praveen2112 commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,70 @@
+        this.delay = delay.toMillis();
+
+        requireNonNull(maxDelay);
+        this.maxDelay = maxDelay.toMillis();
```
Can we inline them with the previous statement ?

takezoe replied:
Fixed.

Praveen2112 commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,84 @@
+        assertEquals(delegate.counter, 3);
+    }
+
+    @Test(expectedExceptions = FailsafeException.class)
```
Is it possible for us to wrap with Presto's JDBC exception ?

takezoe replied:
Which is better?
```java
        try {
            return Failsafe.with(retryPolicy).get(() -> delegate.openConnection(identity));
        }
        catch (FailsafeException e) {
            throw new PrestoException(JDBC_ERROR, e);
        }
```
or
```java
        try {
            return Failsafe.with(retryPolicy).get(() -> delegate.openConnection(identity));
        }
        catch (FailsafeException e) {
            throw new SQLException(e);
        }
```
The original signature of `openConnection()` has `throws SQLException`. So, it might be better to wrap by `SQLException`.

takezoe replied:
I updated to throw `SQLException` because `getConnection` is typically expected to throw it rather than `PrestoException`.

takezoe commented:
Squashed all commits and rebased to the latest master. Remaining thing is how to wrap Exception. Which is better `PrestoException(JDBC_ERROR, e)` or `SQLException(e)`?

takezoe commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryJdbcConfig.java:
```diff
@@ -0,0 +1,59 @@
+                .put("max-retries", "20")
+                .put("retry-delay-factor", "1.5")
+                .put("retry-jitter-factor", "0.1")
+                .build();
```
Adding the same prefix for the configuration keys might be better?
```
connect.retry-delay
connect.max-retry-delay
connect.max-retries
...
```

takezoe replied:
For now, I added `connect.` prefix. Let me know if it's ok or not.

findepi commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+
+        Connection conn = connectionFactory.openConnection(new JdbcIdentity("test", Optional.empty(), ImmutableMap.of()));
+
+        assertNull(conn);
```
It should not be `null`. `openConnection` should either return a Connection or throw if it cannot

takezoe replied:
Fixed test to return mock `Connection`.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,97 @@
+        return jitterFactor;
+    }
+
+    @Config("retry-jitter-factor")
```
Does this need to be configurable?

if yes => add ConfigDescription (it's not "obvious")

but i'd rather remove it

takezoe replied:
Just removed from `RetryJdbcConfig` and updated `RetryableConnectionFactory` back to use `0.2` as a jitter factor.

findepi commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @NotNull
+    @MinDuration("0ms")
```
Why 1s min? (i could want to retry without a delay)

takezoe commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,72 @@
+    public void close()
+            throws SQLException
+    {
+        this.delegate.close();
```
Fixed

takezoe commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,97 @@
+
+public class RetryJdbcConfig
+{
+    private Duration delay = Duration.valueOf("1s");
```
Fixed

findepi commented:
@electrum do you know of mysql driver retries internally by default?
then we should not add another retry on top

takezoe commented:
@Praveen2112 @findepi Thanks for your review! I updated the pull request following your comment!

> do you know of mysql driver retries internally by default? then we should not add another retry on top

At least, PostgreSQL doesn't. So, if we shouldn't add this generally, I will move this to PostgreSQL connector.

colebow commented:
:wave: @takezoe - this PR has become inactive. If you're still interested in working on it, please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.

takezoe commented:
@colebow I'm not sure if this approach is acceptable because no response from reviewers after my last activity. Feel free to close this if no possibility of acceptance, but I will work on resolving conflicts and further enhancements if there is a possibility.

bitsondatadev commented:
Hey @takezoe, I'm gonna ask @Praveen2112 if he has any thoughts on this.

hashhar commented:
@kokosing / @wendigo Will this also get handled by the retries we were looking to put into the SplitManager and PageSource?

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryJdbcConfig.java:
```diff
@@ -0,0 +1,85 @@
+        return delay;
+    }
+
+    @Config("connect.retry-delay")
```
I would move it to `QueryConfig` and use `query` as prefix for config toggles.

kokosing commented on presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/RetryableConnectionFactory.java:
```diff
@@ -0,0 +1,76 @@
+    public Connection openConnection(JdbcIdentity identity) throws SQLException
+    {
+        RetryPolicy<Object> retryPolicy = new RetryPolicy<>()
+                .handle(SQLException.class)
```
it is to simple condition. I think it is very vendor specific and we need to be able inject that condition in guice per each jdbc connector. By default we can retry only on `java.sql.SQLTransientException`

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+    @Test
+    public void testEverythingImplemented()
+    {
+        InterfaceTestUtils.assertAllMethodsOverridden(ConnectionFactory.class, RetryableConnectionFactory.class);
```
static import

kokosing commented on presto-base-jdbc/pom.xml:
```diff
@@ -184,5 +189,12 @@
         </dependency>
+
+        <dependency>
+            <groupId>org.mockito</groupId>
```
please remove

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+        // Max retries = 3 by default
+        ConnectionFactory connectionFactory = new RetryableConnectionFactory(delegate, config);
+
+        Connection conn = connectionFactory.openConnection(new JdbcIdentity("test", Optional.empty(), ImmutableMap.of()));
```
nit: conn -> connection

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    private static class TestConnectionFactory
+            implements ConnectionFactory
```
You don't need a class as `ConnectionFactory` is `FunctionalInterface`. So it could be plain lambda defined in each test.

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+        {
+            counter++;
+            if (counter <= numberOfFailure) {
+                throw new SQLException("Failed to connect");
```
Throw `SQLTransientException`

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+        assertEquals(delegate.counter, 3);
+    }
+
+    @Test(expectedExceptions = SQLException.class)
```
use `assertThatThrownBy` instead

kokosing commented on presto-base-jdbc/src/test/java/io/prestosql/plugin/jdbc/TestRetryableConnectionFactory.java:
```diff
@@ -0,0 +1,91 @@
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+public class TestRetryableConnectionFactory
```
Please test delays too

kokosing commented:
> Will this also get handled by the retries we were looking to put into the SplitManager and PageSource?

This is helpful, but it is also part of the solution. We need to be able to retry queries, however that has done per connector I think.

kokosing commented:
@takezoe Are you willing to continue to work on this? Or would you like someone to take it over?

takezoe commented:
Thanks for comments. I will update this pull request.

takezoe commented:
I realized that there is a similar ConnectionFactory implementation introduced by https://github.com/trinodb/trino/pull/4947. It's currently used only in oracle plugin though enhancing it would be better than adding another one. I would close this pull request and raise new pull request for it.

kokosing commented:
Great idea! 

I think we could make this method https://github.com/trinodb/trino/pull/4947/files#diff-40faa2d5b749c82fbc547507c59b77d293c2a33db0bb3447926e2265e366b392R70 to be part of the `JdbcClient`

takezoe commented:
Created new pull request: https://github.com/trinodb/trino/pull/15205
